src/style-spec/function/index.js

```diff
@@ -51,52 +52,76 @@ type FunctionParameters = DataDrivenPropertyValueSpecification<StylePropertyValu
 function createFunction(parameters: FunctionParameters, propertySpec: StylePropertySpecification): StyleFunction {
     let expr;
 
+    let defaultValue = propertySpec.default;
     if (!isFunctionDefinition(parameters)) {
         expr = convert.value(parameters, propertySpec);
         if (expr === null) {
             expr = getDefaultValue(propertySpec);
         }
     } else if (typeof parameters === 'object' && parameters !== null && typeof parameters.expression !== 'undefined') {
-        expr = ['coalesce', parameters.expression, getDefaultValue(propertySpec)];
+        expr = parameters.expression;
     } else {
         expr = convert.function(parameters, propertySpec);
+        if (parameters && typeof parameters.default !== 'undefined') {
+            defaultValue = parameters.default;
+        }
+    }
+
+    if (propertySpec.type === 'color') {
+        defaultValue = parseColor((defaultValue: any));
     }
 
-    const expectedType = getExpectedType(propertySpec);
-    const compiled = compileExpression(expr, expectedType);
-    if (compiled.result === 'success') {
-        const warningHistory: {[key: string]: boolean} = {};
-        const f = function (globalProperties: {+zoom?: number}, feature?: Feature) {
-            try {
-                const val = compiled.function(globalProperties, feature);
-                return val === null ? undefined : val;
-            } catch (e) {
-                if (!warningHistory[e.message]) {
-                    warningHistory[e.message] = true;
-                    if (typeof console !== 'undefined') console.warn(e.message);
+    let evaluate: StyleFunction;
+
+    if (expr === null) {
+        evaluate = function () { return defaultValue; };
+        evaluate.isFeatureConstant = true;
+        evaluate.isZoomConstant = true;
+    } else {
+        const expectedType = getExpectedType(propertySpec);
+        const compiled = compileExpression(expr, expectedType);
+        if (compiled.result === 'success') {
+            const warningHistory: {[key: string]: boolean} = {};
+            evaluate = function (globalProperties: {+zoom?: number}, feature?: Feature) {
+                try {
+                    const val = compiled.function(globalProperties, feature);
+                    if (val === null || val === undefined) {
+                        return defaultValue;
+                    }
+                    return val;
+                } catch (e) {
```

Comment:
Is there any way we could avoid a full `try/catch` here? Maybe passing an `errors` array to the compiled function and accumulating errors there somehow?

Reply:
Definitely possible, but I think it's outside the scope of this PR -- I experimented with a non-try catch approach recently, and it's definitely a nontrivial change.  The naive approach (checking for errors before evaluating any expression) was slower than try/catch, and the non-naive approach will take some careful work.

Reply:
@anandthakker I'm not sure how much exactly it affects performance (I guess it depends on browser engine/version), but as a rule of thumb, I avoid `arguments` keyword altogether in perf-sensitive code paths. At least it is easier to make sure a certain function is monomorphic.